Add connection tracking metrics in flow exporter

dreamtalen opened the pull request:
Fix #1033
Add there metrics TotalConnectionsInConnTrackTable, TotalAntreaConnectionsInConnTrackTable, MaxConnectionsInConnTrackTable to show the capacity and current number of connection in the connection tracking table.
Add unit tests for these metrics too.
These metrics will be useful for configuring connection tracking table size and also help generate alerts.

srikartati commented on pkg/agent/flowexporter/connections/conntrack_linux.go:
```diff
@@ -44,11 +49,26 @@ func NewConnTrackSystem(nodeConfig *config.NodeConfig, serviceCIDR *net.IPNet) *
 	sysctl.EnsureSysctlNetValue("netfilter/nf_conntrack_timestamp", 1)
 
-	return &connTrackSystem{
+	ct := &connTrackSystem{
```
You don't seem to be using this variable. Why not keep the same return that is previously there?

dreamtalen replied:
Thanks, done.

srikartati commented on pkg/agent/flowexporter/connections/conntrack_ovs.go:
```diff
@@ -49,11 +50,25 @@ func NewConnTrackOvsAppCtl(nodeConfig *config.NodeConfig, serviceCIDR *net.IPNet
 		return nil
 	}
-	return &connTrackOvsCtl{
+	ct := &connTrackOvsCtl{
```
same here

dreamtalen replied:
Variable ct needed here to do RunAppctlCmd call for getting maxconns.

srikartati replied:
can't we use ovsctlClient directly here? Moving the initialization of structure to the end makes the code easy to read I think.

dreamtalen replied:
Make senses. Fixed it, thanks!

srikartati commented on pkg/agent/flowexporter/connections/conntrack_test.go:
```diff
@@ -157,5 +170,26 @@ func TestConnTackOvsAppCtl_DumpFlows(t *testing.T) {
+	checkMaxConnectionsMetrics(t, maxConns)
+}
+
+func checkConnectionMetrics(t *testing.T, numConns int) {
```
checkConnectionMetrics->checkTotalConnectionsMetric

dreamtalen replied:
Thanks, done.

srikartati commented on pkg/agent/flowexporter/connections/conntrack_test.go:
```diff
@@ -157,5 +170,26 @@ func TestConnTackOvsAppCtl_DumpFlows(t *testing.T) {
+	assert.Equal(t, nil, testutil.GatherAndCompare(legacyregistry.DefaultGatherer, strings.NewReader(expectedConnectionCount), "antrea_agent_conntrack_total_connection_count"))
+}
+
+func checkMaxConnectionsMetrics(t *testing.T, maxConns int) {
```
checkMaxConnectionsMetrics -> checkMaxConnectionsMetric

dreamtalen replied:
Thanks, done.

srikartati commented on pkg/agent/flowexporter/connections/conntrack_test.go:
```diff
@@ -157,5 +170,26 @@ func TestConnTackOvsAppCtl_DumpFlows(t *testing.T) {
+	# HELP antrea_agent_conntrack_max_connection_count [ALPHA] Size of the conntrack table.
+	# TYPE antrea_agent_conntrack_max_connection_count gauge
+	`
+	expectedkMaxConnectionsCount = expectedkMaxConnectionsCount + fmt.Sprintf("antrea_agent_conntrack_max_connection_count %d\n", maxConns)
```
typo---expectedk?

dreamtalen replied:
Thanks, done.

srikartati commented on pkg/agent/flowexporter/connections/conntrack_linux.go:
```diff
@@ -44,11 +49,26 @@ func NewConnTrackSystem(nodeConfig *config.NodeConfig, serviceCIDR *net.IPNet) *
 	}
+
+	// Set the antrea_agent_conntrack_max_connection_count prometheus metric
+	cmdOutput, err := exec.Command("cat", "/proc/sys/net/nf_conntrack_max").Output()
```
could you use GetSysctlNet from pkg/agent/util/sysctl?

dreamtalen replied:
Thanks, done.

srikartati commented on pkg/agent/flowexporter/connections/conntrack_test.go:
```diff
@@ -17,24 +17,31 @@
 func TestConnTrackSystem_DumpFlows(t *testing.T) {
 	ctrl := gomock.NewController(t)
 	defer ctrl.Finish()
+	metrics.InitializeConnectionMetrics()
```
Could we initialize connDumperDPSystem using NewConnTrackSystem? This will help test the max connection metric and sysctl calls as well. Moreover, this should be ok as these unit tests are run for linux.
You could add assert for nf_conntrack_acct and nf_conntrack_timestamp.
I know that this is not related to connection metric but helps in improving the unit test coverage.

dreamtalen replied:
Thanks, done.

srikartati reviewed (commented):
Thanks for adding these metrics. I have a few comments.

dreamtalen commented on pkg/agent/flowexporter/connections/conntrack_linux.go:
```diff
@@ -18,13 +18,18 @@ package connections
+	"github.com/vmware-tanzu/antrea/pkg/agent/metrics"
 	"github.com/vmware-tanzu/antrea/pkg/agent/util/sysctl"
+
+	"os/exec"
```
Thanks, done.

antoninbas commented on pkg/agent/flowexporter/connections/conntrack_ovs.go:
```diff
@@ -49,6 +50,19 @@ func NewConnTrackOvsAppCtl(nodeConfig *config.NodeConfig, serviceCIDR *net.IPNet
 		return nil
 	}
+
+	cmdOutput, execErr := ovsctlClient.RunAppctlCmd("dpctl/ct-get-maxconns", false)
+	if execErr != nil {
+		klog.Errorf("error when executing dpctl/ct-get-maxconns command: %v", execErr)
+		return nil
+	}
+	maxConns, err := strconv.Atoi(strings.TrimSpace(string(cmdOutput)))
+	if err != nil {
+		klog.Errorf("error when convert dpctl/ct-get-maxconns cmdoutput to int: %v", err)
+		return nil
+	}
```
place this in helper function

dreamtalen replied:
Thanks, done.

antoninbas commented on pkg/agent/flowexporter/connections/conntrack_ovs.go:
```diff
@@ -49,6 +50,19 @@ func NewConnTrackOvsAppCtl(nodeConfig *config.NodeConfig, serviceCIDR *net.IPNet
+	}
+	maxConns, err := strconv.Atoi(strings.TrimSpace(string(cmdOutput)))
+	if err != nil {
+		klog.Errorf("error when convert dpctl/ct-get-maxconns cmdoutput to int: %v", err)
```
s/convert/converting

dreamtalen replied:
Thanks, done.

antoninbas commented on pkg/agent/flowexporter/connections/connections_test.go:
```diff
@@ -286,5 +297,16 @@ func TestConnectionStore_DeleteConnectionByKey(t *testing.T) {
+	`
+	expectedAntreaConnectionCount = expectedAntreaConnectionCount + fmt.Sprintf("antrea_agent_conntrack_antrea_connection_count %d\n", numConns)
+
+	assert.Equal(t, nil, testutil.GatherAndCompare(legacyregistry.DefaultGatherer, strings.NewReader(expectedAntreaConnectionCount), "antrea_agent_conntrack_antrea_connection_count"))
```
`assert.NoError`

dreamtalen replied:
Thanks, I also fixed it for other 2 asser.Equal() call.

antoninbas commented on pkg/agent/flowexporter/connections/conntrack_linux.go:
```diff
@@ -44,6 +45,14 @@ func NewConnTrackSystem(nodeConfig *config.NodeConfig, serviceCIDR *net.IPNet) *
 	// Do not fail, but continue after logging error as we can still dump flows with no timestamps.
 	sysctl.EnsureSysctlNetValue("netfilter/nf_conntrack_timestamp", 1)
 
+	// Set the antrea_agent_conntrack_max_connection_count prometheus metric
+	maxConns, err := sysctl.GetSysctlNet("nf_conntrack_max")
+	if err != nil {
+		klog.Errorf("error when getting nf_conntrack_max: %v", err)
+		return nil
+	}
+	metrics.MaxConnectionsInConnTrackTable.Set(float64(maxConns))
```
I assume this can vary dynamically while the agent is running. The admin is free to update this value, so maybe we should have a background task to keep it up-to-date? We should have a test for that too if possible.

antoninbas commented on pkg/agent/flowexporter/connections/conntrack_ovs.go:
```diff
@@ -49,6 +50,19 @@ func NewConnTrackOvsAppCtl(nodeConfig *config.NodeConfig, serviceCIDR *net.IPNet
+		klog.Errorf("error when convert dpctl/ct-get-maxconns cmdoutput to int: %v", err)
+		return nil
+	}
+	metrics.MaxConnectionsInConnTrackTable.Set(float64(maxConns))
```
here also I assume that this may change dynamically (see comment above for Linux implementation)? maybe we should have an interface method to read the max number of connections (implemented differently based on the platform) and have one background task to update that metric. Or maybe we just refresh the metric every time we read the connections: it may be the simplest solution and my favorite one :) @srikartati

srikartati replied:
Thats a good point @antoninbas. Agree that this number can be changed by admin and we should track this differently. Yes reading max conn metric when reading the connection should be a simple solution. In addition, we should add in all the metrics help statement saying that these metrics are updated every pollInterval. As sometimes depending on poll interval and scrape interval of prometheus server, metrics may not be updated in prometheus server.

dreamtalen replied:
Thanks, I moved the updating max connection logics inside the DumpFlows() function as a simple solution right now.

antoninbas commented on pkg/agent/flowexporter/connections/conntrack_ovs.go:
```diff
@@ -201,3 +213,15 @@ func lookupProtocolMap(name string) (uint8, error) {
+	}
+	maxConns, err := strconv.Atoi(strings.TrimSpace(string(cmdOutput)))
+	if err != nil {
+		return 0, fmt.Errorf("error when converting dpctl/ct-get-maxconns cmdoutput %s to int", cmdOutput)
```
```suggestion
		return 0, fmt.Errorf("error when converting dpctl/ct-get-maxconns output '%s' to int", cmdOutput)
```

dreamtalen replied:
Thanks, done.

antoninbas commented on pkg/agent/flowexporter/connections/conntrack_linux.go:
```diff
@@ -68,6 +69,17 @@ func (ct *connTrackSystem) DumpFlows(zoneFilter uint16) ([]*flowexporter.Connect
 		return nil, err
 	}
+
+	// Update the antrea_agent_conntrack_max_connection_count prometheus metric
```
s/prometheus/Prometheus

dreamtalen replied:
Thanks, done.

antoninbas commented on pkg/agent/flowexporter/connections/conntrack_linux.go:
```diff
@@ -68,6 +69,17 @@ func (ct *connTrackSystem) DumpFlows(zoneFilter uint16) ([]*flowexporter.Connect
 	}
+
+	// Update the antrea_agent_conntrack_max_connection_count prometheus metric
+	maxConns, err := sysctl.GetSysctlNet("nf_conntrack_max")
+	if err != nil {
+		klog.Errorf("Error when getting nf_conntrack_max: %v", err)
+		return nil, err
+	}
```
I suggest the following for symmetry with the userspace case and so that we get a bit more context in the log message:
```go
getMaxConnectionsInConnTrack := func() (int, err) {
    maxConns, err := sysctl.GetSysctlNet("nf_conntrack_max")
    if err != nil {
        return 0, fmt.Errorf("rrror when getting nf_conntrack_max: %v", err)
    }
    return maxConns, nil
}

maxConns, err := getMaxConnectionsInConnTrack()
if err != nil {
    klog.Errorf("Error when getting max connections in conntrack: %v", err)
    return nil, err
}
```

dreamtalen replied:
Thanks, done.

antoninbas commented on pkg/agent/flowexporter/connections/conntrack_linux.go:
```diff
@@ -68,6 +69,17 @@ func (ct *connTrackSystem) DumpFlows(zoneFilter uint16) ([]*flowexporter.Connect
+		klog.Errorf("Error when getting nf_conntrack_max: %v", err)
+		return nil, err
+	}
+	metrics.MaxConnectionsInConnTrackTable.Set(float64(maxConns))
+
+	metrics.TotalConnectionsInConnTrackTable.Set(float64(len(conns)))
```
I was wondering if we could avoid updating metrics in platform-dependent code altogether, by adding a new return value to the `DumpFlows` method in the `ConnTrackDumper` interface:
```go
type conntrackOccupancy struct {
    maxConnections int
    totalConnections int
}

type ConnTrackDumper interface {
    DumpFlows(zoneFilter uint16) ([]*flowexporter.Connection, conntrackOccupancy, error)
}
```

then we can update the metrics in platform-independent code (connections.go)

just an idea, please let me know what you think

dreamtalen replied:
Thanks, it's a good idea especially considering we will add Windows support for flow exporter in the future. I rewrite the unit tests too after changing the interface in this way.

antoninbas commented on pkg/agent/flowexporter/connections/conntrack_ovs.go:
```diff
@@ -64,6 +66,14 @@ func (ct *connTrackOvsCtl) DumpFlows(zoneFilter uint16) ([]*flowexporter.Connect
 		return nil, err
 	}
 
+	// Update the antrea_agent_conntrack_max_connection_count prometheus metric
```
s/prometheus/Prometheus

antoninbas reviewed (commented):
Thanks for addressing my earlier comments, I added a couple more

srikartati commented on pkg/agent/flowexporter/connections/conntrack_linux.go:
```diff
@@ -66,12 +67,27 @@ func (ct *connTrackSystem) DumpFlows(zoneFilter uint16) ([]*flowexporter.Connect
+		}
+		return maxConns, nil
 	}
+	// Update the antrea_agent_conntrack_max_connection_count Prometheus metric value
```
This comment has to be changed now.

dreamtalen replied:
Thanks, done.

srikartati commented on pkg/agent/flowexporter/connections/connections.go:
```diff
@@ -190,7 +192,7 @@ func (cs *ConnectionStore) Poll() (int, error) {
 	cs.ForAllConnectionsDo(resetConn)
 
-	filteredConns, err := cs.connDumper.DumpFlows(openflow.CtZone)
+	filteredConns, conntrackOccupancyMetrics, err := cs.connDumper.DumpFlows(openflow.CtZone)
```
same here.. a shorter variable name is better.

dreamtalen replied:
Thanks, done.

srikartati commented on pkg/agent/flowexporter/connections/connections_test.go:
```diff
@@ -286,5 +299,97 @@ func TestConnectionStore_DeleteConnectionByKey(t *testing.T) {
+	// Reference for flow timestamp
+	refTime := time.Now()
+	// Create two flows; one is already in ConnectionStore and other one is new
+	testFlows := make([]*flowexporter.Connection, 2)
```
do you need these test flows? We do not have to test addOrUpdateConn here. You can consider `filteredConns` returned by DumpFlows as empty and just test Prometheus metrics.

dreamtalen replied:
Make sense, use a empty test flows list instead.

dreamtalen commented on pkg/agent/flowexporter/connections/conntrack_ovs.go:
```diff
@@ -57,24 +58,31 @@ func NewConnTrackOvsAppCtl(nodeConfig *config.NodeConfig, serviceCIDR *net.IPNet
+		return nil, nil, err
+	}
+
+	// Update the antrea_agent_conntrack_max_connection_count Prometheus metric value
```
Thanks, done.

dreamtalen commented on pkg/agent/flowexporter/connections/conntrack_linux.go:
```diff
@@ -66,12 +67,27 @@ func (ct *connTrackSystem) DumpFlows(zoneFilter uint16) ([]*flowexporter.Connect
 	klog.V(2).Infof("No. of flow exporter considered flows in Antrea zoneID: %d", len(filteredConns))
 
-	return filteredConns, nil
+	return filteredConns, &flowexporter.ConntrackOccupancy{maxConns, len(conns)}, nil
```
Thanks, done.

dreamtalen commented on pkg/agent/flowexporter/connections/conntrack_test.go:
```diff
@@ -151,11 +174,12 @@ func TestConnTackOvsAppCtl_DumpFlows(t *testing.T) {
 	mockOVSCtlClient.EXPECT().RunAppctlCmd("dpctl/dump-conntrack", false, "-m", "-s").Return(ovsctlCmdOutput, nil)
 
-	conns, err := connDumper.DumpFlows(uint16(openflow.CtZone))
+	conns, conntrackOccupancyMetrics, err := connDumper.DumpFlows(uint16(openflow.CtZone))
```
Thanks, done.

dreamtalen commented on pkg/agent/flowexporter/connections/conntrack_ovs.go:
```diff
@@ -57,24 +58,31 @@ func NewConnTrackOvsAppCtl(nodeConfig *config.NodeConfig, serviceCIDR *net.IPNet
 	klog.V(2).Infof("FlowExporter considered flows: %d", len(filteredConns))
 
-	return filteredConns, nil
+	return filteredConns, &flowexporter.ConntrackOccupancy{maxConns, totalConns}, nil
```
Thanks, done.

srikartati commented on pkg/agent/metrics/prometheus.go:
```diff
@@ -94,6 +94,30 @@ var (
+	TotalConnectionsInConnTrackTable = metrics.NewGauge(
+		&metrics.GaugeOpts{
+			Name:           "antrea_agent_conntrack_total_connection_count",
+			Help:           "Number of connections in the conntrack table, update every flowPollInterval.",
```
suggestion: Number of connections in the conntrack table. This metric gets updated at an interval specified by flowPollInterval, a configMap parameter in antrea-agent.conf.

dreamtalen replied:
Thanks, done.

srikartati commented on pkg/agent/metrics/prometheus.go:
```diff
@@ -94,6 +94,30 @@ var (
+	TotalAntreaConnectionsInConnTrackTable = metrics.NewGauge(
+		&metrics.GaugeOpts{
+			Name:           "antrea_agent_conntrack_antrea_connection_count",
+			Help:           "Number of connections in the Antrea ZoneID of the conntrack table.",
```
Above flowPollInterval statement is applicable here too.

dreamtalen replied:
Antrea connection count is updated when add or delete a connection, so I don't think the flowPollInterval statement applies here.

srikartati replied:
Add and delete connection logic is only executed through connectionStore.Poll which is called at a given flowPollInterval, so the metric update is dependent on it.

dreamtalen replied:
Got it, thanks!

srikartati commented on pkg/agent/metrics/prometheus.go:
```diff
@@ -94,6 +94,30 @@ var (
+	MaxConnectionsInConnTrackTable = metrics.NewGauge(
+		&metrics.GaugeOpts{
+			Name:           "antrea_agent_conntrack_max_connection_count",
+			Help:           "Size of the conntrack table, update every flowPollInterval.",
```
same here.

dreamtalen replied:
Thanks, done.

antoninbas reviewed (commented):
Looks good on my side. Ultimately, I wonder if we will want to decouple the conntrack occupancy metrics from the flow exporter implementation. @srikartati is it possible to enable connection polling by providing a flowPollInterval value, without enabling flow export (i.e. without a collector). If not, do you think this is something we could have in a future PR?

srikartati commented:
> Looks good on my side. Ultimately, I wonder if we will want to decouple the conntrack occupancy metrics from the flow exporter implementation. @srikartati is it possible to enable connection polling by providing a flowPollInterval value, without enabling flow export (i.e. without a collector). If not, do you think this is something we could have in a future PR?

@antoninbas Currently, no it is not straightforward to decouple both polling and flow export mechanisms. They are both enabled through FlowExporter feature gate and we expect flowCollectorAddr as a must given parameter. Yes, it is possible to decouple, but we probably have to either move polling from the FlowExporter feature gate or change the name of the feature gate to a more appropriate name and remove the dependency on flowCollectorAddr. We should definitely pursue this as a future PR. 
@dreamtalen Yongming, could you open an issue for this, so we will not lose track of this?

srikartati commented:
@dreamtalen There is a unit test failure: https://github.com/vmware-tanzu/antrea/pull/1232/checks?check_run_id=1131918989 Please take a look.

dreamtalen commented:
> > Looks good on my side. Ultimately, I wonder if we will want to decouple the conntrack occupancy metrics from the flow exporter implementation. @srikartati is it possible to enable connection polling by providing a flowPollInterval value, without enabling flow export (i.e. without a collector). If not, do you think this is something we could have in a future PR?
> 
> @antoninbas Currently, no it is not straightforward to decouple both polling and flow export mechanisms. They are both enabled through FlowExporter feature gate and we expect flowCollectorAddr as a must given parameter. Yes, it is possible to decouple, but we probably have to either move polling from the FlowExporter feature gate or change the name of the feature gate to a more appropriate name and remove the dependency on flowCollectorAddr. We should definitely pursue this as a future PR.
> @dreamtalen Yongming, could you open an issue for this, so we will not lose track of this?

dreamtalen commented:
> > Looks good on my side. Ultimately, I wonder if we will want to decouple the conntrack occupancy metrics from the flow exporter implementation. @srikartati is it possible to enable connection polling by providing a flowPollInterval value, without enabling flow export (i.e. without a collector). If not, do you think this is something we could have in a future PR?
> 
> @antoninbas Currently, no it is not straightforward to decouple both polling and flow export mechanisms. They are both enabled through FlowExporter feature gate and we expect flowCollectorAddr as a must given parameter. Yes, it is possible to decouple, but we probably have to either move polling from the FlowExporter feature gate or change the name of the feature gate to a more appropriate name and remove the dependency on flowCollectorAddr. We should definitely pursue this as a future PR.
> @dreamtalen Yongming, could you open an issue for this, so we will not lose track of this?

Sure

antoninbas commented on pkg/agent/flowexporter/connections/conntrack_linux.go:
```diff
@@ -66,12 +67,27 @@ func (ct *connTrackSystem) DumpFlows(zoneFilter uint16) ([]*flowexporter.Connect
+	getMaxConnectionsInConnTrack := func() (int, error) {
+		maxConns, err := sysctl.GetSysctlNet("nf_conntrack_max")
+		if err != nil {
+			return 0, fmt.Errorf("Error when getting nf_conntrack_max: %v", err)
```
nit: we do not capitalize error messages when using `fmt.Error`, but error logs (`klog.Error`) should be capitalized

dreamtalen replied:
Got it, thanks!

antoninbas reviewed (approved):
LGTM, just one nit. @srikartati please merge when you're satisfied with the changes.

dreamtalen commented:
> @dreamtalen There is a unit test failure: https://github.com/vmware-tanzu/antrea/pull/1232/checks?check_run_id=1131918989 Please take a look.

Fixed, this unit test fails because the kind cluster doesn't have privilege to set the nf_conntrack_acct and nf_conntrack_timestamp. 
Added error return value for NewConnTrackSystem() and bypass the assert statement if the sysctl setting failed

antoninbas commented on pkg/agent/flowexporter/connections/conntrack_test.go:
```diff
@@ -76,28 +82,59 @@ func TestConnTrackSystem_DumpFlows(t *testing.T) {
 	assert.Equal(t, 1, len(conns), "number of filtered connections should be equal")
+
+	if acctErr == nil {
+		conntrackAcct, err := sysctl.GetSysctlNet("netfilter/nf_conntrack_acct")
```
@dreamtalen @srikartati Sorry I missed that earlier. A unit test should ideally not rely on platform-specific constructs. Is it possible to move this to integration tests? Or define an interface to access that information?

srikartati replied:
@antoninbas I made this platform-dependent by adding // +build linux on top as we only support linux for now. Took the cue from pkg/agent/route/route_windows_test.go

I see the point of making the unit tests independent of platforms. 
For now, could we change the name of file like above: conntrack_linux_test.go?
When the windows support comes in the imminent future, we modify the file with interfaces for sysctl calls that works for all platforms to make these unit tests platform-independent in future PR?

antoninbas commented:
> > @dreamtalen There is a unit test failure: https://github.com/vmware-tanzu/antrea/pull/1232/checks?check_run_id=1131918989 Please take a look.
> 
> Fixed, this unit test fails because the kind cluster doesn't have privilege to set the nf_conntrack_acct and nf_conntrack_timestamp.
> Added error return value for NewConnTrackSystem() and bypass the assert statement if the sysctl setting failed

We do not run unit tests in Kind cluster...

srikartati commented on pkg/agent/flowexporter/connections/conntrack_test.go:
```diff
@@ -76,28 +82,59 @@ func TestConnTrackSystem_DumpFlows(t *testing.T) {
+
+	if acctErr == nil {
+		conntrackAcct, err := sysctl.GetSysctlNet("netfilter/nf_conntrack_acct")
+		// Ignore the permission denied error if test cluster doesn't have privilege
```
nit: test platform is appropriate as this is just unit test.

dreamtalen commented:
> > > @dreamtalen There is a unit test failure: https://github.com/vmware-tanzu/antrea/pull/1232/checks?check_run_id=1131918989 Please take a look.
> > 
> > 
> > Fixed, this unit test fails because the kind cluster doesn't have privilege to set the nf_conntrack_acct and nf_conntrack_timestamp.
> > Added error return value for NewConnTrackSystem() and bypass the assert statement if the sysctl setting failed
> 
> We do not run unit tests in Kind cluster...
Let me check with Srikar again, because he told me the unit tests are running on the github Kind cluster so we did this fix

srikartati commented on pkg/agent/flowexporter/connections/conntrack_test.go:
```diff
@@ -76,28 +82,59 @@ func TestConnTrackSystem_DumpFlows(t *testing.T) {
 	}
 	assert.Equal(t, 1, len(conns), "number of filtered connections should be equal")
+
+	if acctErr == nil {
+		conntrackAcct, err := sysctl.GetSysctlNet("netfilter/nf_conntrack_acct")
+		// Ignore the permission denied error if test cluster doesn't have privilege
+		if err != nil && !os.IsPermission(err) {
+			t.Errorf("error when getting nf_conntrack_acct: %v", err)
+		} else if err == nil {
+			assert.Equal(t, 1, conntrackAcct, "net.netfilter.nf_conntrack_acct value should be 1")
+		}
+	}
+
+	if timestampErr == nil {
+		conntrackTimestamping, err := sysctl.GetSysctlNet("netfilter/nf_conntrack_timestamp")
+		// Ignore the permission denied error if test cluster doesn't have privilege
+		if err != nil && !os.IsPermission(err) {
+			t.Errorf("error when getting nf_conntrack_timestamp: %v", err)
+		} else if err == nil {
+			assert.Equal(t, 1, conntrackTimestamping, "net.netfilter.nf_conntrack_timestamp value should be 1")
+		}
+	}
```
I think moving these lines to after L85 makes more positional sense.

srikartati reviewed (commented):
LGTM except for a couple of nits. Thanks for fixing the unit test.

srikartati commented:
> > > Fixed, this unit test fails because the kind cluster doesn't have privilege to set the nf_conntrack_acct and nf_conntrack_timestamp.
> > > Added error return value for NewConnTrackSystem() and bypass the assert statement if the sysctl setting failed
> > 
> > 
> > We do not run unit tests in Kind cluster...
> 
> Let me check with Srikar again, because he told me the unit tests are running on the github Kind cluster so we did this fix.
> 
> Checked, miscommunication here, the github CI platform running unit test doesn't have privilege too, like the kind cluster.

I think there was communication gap. I mentioned that there are privilege issues when running e2e tests in kind clusters. Github CI workers run units.

srikartati commented:
/test-all

antoninbas commented on pkg/agent/flowexporter/connections/conntrack_test.go:
```diff
@@ -76,28 +82,59 @@ func TestConnTrackSystem_DumpFlows(t *testing.T) {
+	}
+
+	maxConns, err := sysctl.GetSysctlNet("nf_conntrack_max")
+	// Ignore the permission denied error if test cluster doesn't have privilege
```
The comment is misleading: these tests are not run in clusters. The Go unit tests are run on Linux & Windows VMs (well these ones only on Linux because of the build tag). Unit tests should ideally never require special privileges. So right now in CI, we will never be able to exercise the normal code path, because the sysctl stuff will always fail.

dreamtalen replied:
Thanks, I will modify the comment. By the way, the Github CI running unit tests didn't fail on the sysctl, so this if condition is used to bypass the asset statement if the sysctl command failed because of the permission error on other CI platforms.

antoninbas replied:
I thought that these checks were introduced because a CI job was failing originally?

BTW my earlier comment is wrong: it seems we only run unit tests on Linux, not macOS

dreamtalen replied:
Yes. Actually, the NewConnTrackSystem function meets the permission error when setting nf_conntrack_acct and nf_conntrack_timestamp and these error logs have been there before. Since we add assert statements for these two values in this unit test, the CI job begins to fail because test VM could get these values using sysctl command so there is a mismatch. In the lastest commit, we fix this failure by adding the error return value of NewConnTrackSystem, if the setting process failed we will not do the assert statement. But since I'm not sure if the GetSysctlNet will fail or not when setting process succeeded, so I add the if condition for GetSysctlNet too.

srikartati commented on test/integration/agent/flowexporter_test.go:
```diff
@@ -127,8 +127,13 @@ func TestConnectionStoreAndFlowRecords(t *testing.T) {
+	}
 	// Expect calls for connStore.poll and other callees
-	connDumperMock.EXPECT().DumpFlows(uint16(openflow.CtZone)).Return(testConns, nil)
+	connDumperMock.EXPECT().DumpFlows(uint16(openflow.CtZone)).Return(testConns, testConnectionMetric, nil)
```
can't you pass nil here? testConnectionMetric is not used anywhere right? It is better to avoid as there is no metric test here.

dreamtalen replied:
Thanks, done.

srikartati reviewed (commented):
Could you change the name of conntrack_test.go -> conntrack_linux_test.go to make dependency on Linux explicit?
Please remove the dependency on the linux platform when adding flow exporter support for windows nodes. 

Once these changes are made, we can check in the code, if there are no other comments from @antoninbas.

srikartati commented:
/test-all

antoninbas commented on pkg/agent/flowexporter/connections/conntrack_linux.go:
```diff
@@ -36,28 +37,28 @@ type connTrackSystem struct {
 }
 
-func NewConnTrackSystem(nodeConfig *config.NodeConfig, serviceCIDR *net.IPNet) *connTrackSystem {
+func NewConnTrackSystem(nodeConfig *config.NodeConfig, serviceCIDR *net.IPNet) (*connTrackSystem, error, error) {
```
it makes no sense to me that the function is returning 2 errors, that's hardly a good design IMO

and the needs of unit tests don't really justify that change

dreamtalen replied:
Fixed. Declared a new function called setupConntrackParameters(), returns the first error encountered since setting nf_conntrack_acct and nf_conntrack_timestamp will succeed or fail at the same time.

antoninbas reviewed (commented):
I think we can still merge it for 0.10 and improve the unit tests as we add Windows support.

However, I am still waiting for an answer as to whether there is *any* CI job that will exercise the gated asserts. AFAIK, there is none, so the only way to exercise that code path is to run the test locally as a privileged user, which is not how I would typically run unit tests.

srikartati commented on pkg/agent/flowexporter/connections/conntrack_linux.go:
```diff
@@ -66,12 +67,27 @@ func (ct *connTrackSystem) DumpFlows(zoneFilter uint16) ([]*flowexporter.Connect
+
+	maxConns, err := getMaxConnectionsInConnTrack()
+	if err != nil {
+		klog.Errorf("Error when getting max connections in conntrack: %v", err)
```
Yes, returning error is good enough as we log the error when there in connectionStore.Run with call to connectionStore.Poll. Please just return the error here and at L69, and do not log it. Is this what you meant @antoninbas ?

In addition, I feel one message is enough like below rather than two messages "error when getting max connections through nf_conntrack_max" when logging.

antoninbas replied:
Yes that's what I meant. You can just return the error here, it will be logged by the caller

dreamtalen replied:
Got it, thanks. I find simillar situations at L60, L69 and L64, L70 in conntrack_ovs.go, shall I remove all these log messages and return the error instead?

antoninbas replied:
Yes: if you find yourself logging and returning the same error, you probably want to remove the log (after checking that someone up in the call chain will log the error). In some rare cases I guess you may want an extra log if it provides additional information.

srikartati commented on pkg/agent/flowexporter/connections/conntrack_linux_test.go:
```diff
@@ -76,28 +82,58 @@ func TestConnTrackSystem_DumpFlows(t *testing.T) {
 	mockNetlinkCT.EXPECT().DumpFilter(conntrack.Filter{}).Return(testFlows, nil)
 
-	conns, err := connDumperDPSystem.DumpFlows(openflow.CtZone)
+	conns, conntrackMetrics, err := connDumperDPSystem.DumpFlows(openflow.CtZone)
 	if err != nil {
 		t.Errorf("Dump flows function returned error: %v", err)
 	}
 	assert.Equal(t, 1, len(conns), "number of filtered connections should be equal")
+
+	maxConns, err := sysctl.GetSysctlNet("nf_conntrack_max")
+	// Bypass this assert statement if the test VM meets a permission error for the sysctl command.
+	if err != nil && !os.IsPermission(err) {
+		t.Errorf("error when getting nf_conntrack_max: %v", err)
+	} else if err == nil {
+		assert.Equal(t, maxConns, conntrackMetrics.MaxConnections, "Size of the conntrack table should be equal to the nf_conntrack_max")
```
I think the issue with Github CI is only write access to sysctl paramters. Read access works fine and we will not get the error in `DumpFlows`. This assert gate will be exercised. Please correct me if I am wrong @dreamtalen.

In the case of conntrack_acct and conntrack_timestamp, yes assert gate is not exercised but at least calls to GetSysCtl code is tested through github CI.

antoninbas replied:
I think this part of the code can be simplified to:
```bash
maxConns, err := sysctl.GetSysctlNet("nf_conntrack_max")
require.NoError(t, err, "Cannot read nf_conntrack_max")
assert.Equal(t, maxConns, conntrackMetrics.MaxConnections, "Size of the conntrack table should be equal to nf_conntrack_max")
```

dreamtalen replied:
Thanks, I also applied this simplification for nf_conntrack_acct and nf_conntrack_timestamp.

srikartati reviewed (commented):
As mentioned in one of the comments, assert gate is executed for nf_conntrack_max but not for conntrack_acct and conntrack_timestamp in github CI. Even though in the case of github CI, assert gate is not executed for the latter, we can get GetSysCtl code being exercised for that part giving some coverage. 

@antoninbas If you feel strongly about not catching the changes in the code in local runs and that is not a strong use-case, we could remove the assert gate for conntrack_acct and conntrack_timestamp and keep it for nf_conntrack_max. I do run unit tests locally as part of dev testing :)

antoninbas commented:
We can keep the conntrack_acct and conntrack_timestamp stuff as is for now and address it later.
You run the unit tests locally but do you run them as root :) ?

antoninbas commented on pkg/agent/flowexporter/connections/conntrack_linux.go:
```diff
@@ -150,3 +154,19 @@ func netlinkFlowToAntreaConnection(conn *conntrack.Flow) *flowexporter.Connectio
 	return &newConn
 }
+
+func setupConntrackParameters() error {
+	// Ensure net.netfilter.nf_conntrack_acct value to be 1. This will enable flow exporter to export stats of connections.
+	// Do not fail, but continue after logging error as we can still dump flows with no stats.
+	acctErr := sysctl.EnsureSysctlNetValue("netfilter/nf_conntrack_acct", 1)
+	if acctErr != nil {
+		return acctErr
+	}
+	// Ensure net.netfilter.nf_conntrack_timestamp value to be 1. This will enable flow exporter to export timestamps of connections.
+	// Do not fail, but continue after logging error as we can still dump flows with no timestamps.
+	timestampErr := sysctl.EnsureSysctlNetValue("netfilter/nf_conntrack_timestamp", 1)
+	if timestampErr != nil {
+		return timestampErr
+	}
+	return nil
+}
```
@srikartati the comments no longer seem to match the implementation. What are your suggestions?

I feel like this got a bit messy because of testing. What we could do is:

 * change the implementation of this function

```go
func setupConntrackParameters() error {
    parametersWithErrors := []string{}
    if sysctl.EnsureSysctlNetValue("netfilter/nf_conntrack_acct", 1) != nil {
        parametersWithErrors = append(parametersWithErrors, "net.netfilter.nf_conntrack_acct")
    }
    if sysctl.EnsureSysctlNetValue("netfilter/nf_conntrack_timestamp", 1) != nil {
        parametersWithErrors = append(parametersWithErrors, "net.netfilter.nf_conntrack_timestamp")
    }
    if len(parametersWithErrors) > 0 {
        return fmt.Errorf("the following kernel parameters could not be verified / set: %v", parametersWithErrors)
    }
    return nil
}
```

 * change the implementation of `NewConnTrackSystem`

```go
func NewConnTrackSystem(nodeConfig *config.NodeConfig, serviceCIDR *net.IPNet) *connTrackSystem {
    if err := setupConntrackParameters(); err != nil {
        // Do not fail, but continue after logging an error as we can still dump flows with missing information.
        klog.Errorf("Error when setting up conntrack parameters, some information may be missing from exported flows: %v", err)
    }
    return &connTrackSystem{...}
}
```

 * in the unit tests: call `setupConntrackParameters()` separately to determine whether these parameters can be set and therefore verified. This is a temporary solution since we will update this code in the future.

srikartati replied:
Yes, I was thinking the same about the old comments.
Testing setupConntrackParameters separately is better in unit tests, and we could create the new object using &connTrackSystem{...} like before in unit tests. 

Pointer 1 is a good idea. Thanks for suggesting.

dreamtalen replied:
Thanks for your advices. I finished changing and added a unit test for setupConntrackParameters() function.

tnqn commented on pkg/agent/flowexporter/connections/conntrack_linux.go:
```diff
@@ -52,26 +50,34 @@ func NewConnTrackSystem(nodeConfig *config.NodeConfig, serviceCIDR *net.IPNet) *
+		return sysctl.GetSysctlNet("nf_conntrack_max")
+	}
+
+	maxConns, err := getMaxConnectionsInConnTrack()
```
why not calling `sysctl.GetSysctlNet("nf_conntrack_max")` directly but creating the one line function?
And getting nf_conntrack_max seems not really related to the function, perhaps the function could return the filtered connections and all connections instead and use another method like `GetMaxConnections` to get `nf_conntrack_max` so that `DumpFlows` can focus on what it desires and no need to create a new struct `ConntrackOccupancy` just to pass the return value.

antoninbas replied:
I agree for the one line function. I think I suggested the helper function back when the function had some more logic.

Regarding adding a new `GetMaxConnections` method to the `ConnTrackDumper` interface, this is also my preferred solution. I suggested returning the values in `DumpFlows` to avoid updating metrics in multiple-places / platform-dependent code (https://github.com/vmware-tanzu/antrea/pull/1232#discussion_r488945704), but Quan's proposal is a nice improvement.

dreamtalen replied:
Hi Quan, as @antoninbas mentioned before, creating a function is for symmetry with the userspace case. For nf_conntrack_max, since this value can be modify by admin anytime, so we decide to read max conn metric when reading the connection as a simple solution right now.

dreamtalen replied:
Thanks for your advices, I have added GetMaxConnections as a new method for ConnTrackDumper interface and added unit tests for GetMaxConnections in kernel and user datapath respectively.

tnqn commented on pkg/agent/flowexporter/connections/conntrack_ovs.go:
```diff
@@ -57,24 +58,30 @@ func NewConnTrackOvsAppCtl(nodeConfig *config.NodeConfig, serviceCIDR *net.IPNet
+		return nil, nil, err
+	}
+
+	maxConns, err := getMaxConnectionsInConnTrack(ct.ovsctlClient)
```
ditto

tnqn commented on pkg/agent/flowexporter/connections/conntrack_ovs.go:
```diff
@@ -201,3 +209,15 @@ func lookupProtocolMap(name string) (uint8, error) {
 	return proto, nil
 }
+
+func getMaxConnectionsInConnTrack(ovsctlClient ovsctl.OVSCtlClient) (int, error) {
```
if the function needs `ovsctlClient`, it makes more sense to be a method of `connTrackOvsCtl`, it's not really different from `ovsAppctlDumpConnections` which also needs `ovsctlClient` and parses the output.

tnqn commented on pkg/agent/metrics/prometheus.go:
```diff
@@ -94,6 +94,30 @@ var (
+	TotalAntreaConnectionsInConnTrackTable = metrics.NewGauge(
+		&metrics.GaugeOpts{
+			Name:           "antrea_agent_conntrack_antrea_connection_count",
+			Help:           "Number of connections in the Antrea ZoneID of the conntrack table. This metric gets updated at an interval specified by flowPollInterval, a configMap parameter in antrea-agent.conf.",
```
I think the metrics is not updated at an interval but updated realtime?

srikartati replied:
As the scraping interval of Prometheus server may not see the changes in the metrics, I want to be explicit about it being updated by specified interval.

antoninbas replied:
it still happens as part of the polling process, but I see your point as we call `Inc` / `Dec` for each inserted / deleted connection, and we don't set the metric value once during polling.

BTW @srikartati do you know why we are not just calling `Set` with `len(filteredConns)`?

srikartati replied:
That is an alternative I thought of but felt that the `Inc/Dec` is a little bit efficient as the number of `Set` calls would be typically higher as we call it every poll interval compared to the case of addition or deletion of the connection.

tnqn commented on pkg/agent/metrics/prometheus.go:
```diff
@@ -94,6 +94,30 @@ var (
+	TotalConnectionsInConnTrackTable = metrics.NewGauge(
+		&metrics.GaugeOpts{
+			Name:           "antrea_agent_conntrack_total_connection_count",
+			Help:           "Number of connections in the conntrack table. This metric gets updated at an interval specified by flowPollInterval, a configMap parameter in antrea-agent.conf.",
```
Not sure `a configMap parameter in antrea-agent.conf` is accurate, maybe `a parameter of antrea-agent.conf in the ConfigMap`? @antoninbas @srikartati

antoninbas replied:
I would just recommend "a configuration parameter for the Agent".

srikartati replied:
Ok with removing antrea-agent.conf

srikartati commented:
LGTM
Thanks for adding changes for prometheus e2e test and fixing comments.

srikartati commented:
/test-all

antoninbas commented on pkg/agent/flowexporter/connections/conntrack_linux_test.go:
```diff
@@ -151,11 +155,47 @@ func TestConnTackOvsAppCtl_DumpFlows(t *testing.T) {
+func TestSetupConnTrackParameters(t *testing.T) {
+	err := setupConntrackParameters()
+	if err != nil {
+		t.Logf("Setup conntrack parameters function returned error: %v", err)
```
how about using `t.Skipf("skipping test as trying to set up conntrack parameters returned an error: %v", err)` ?

dreamtalen replied:
Sounds good, done.

antoninbas commented on pkg/agent/flowexporter/connections/interface.go:
```diff
@@ -22,5 +22,6 @@ import (
 // In future, support will be extended to Windows.
 type ConnTrackDumper interface {
-	DumpFlows(zoneFilter uint16) ([]*flowexporter.Connection, error)
+	DumpFlows(zoneFilter uint16) ([]*flowexporter.Connection, int, error)
+	GetMaxConnections() (int, error)
```
Interface methods should have a comment with a description of the returned parameters

dreamtalen replied:
Thanks, done.

antoninbas commented on pkg/agent/flowexporter/connections/connections.go:
```diff
@@ -190,7 +192,7 @@ func (cs *ConnectionStore) Poll() (int, error) {
 	cs.ForAllConnectionsDo(resetConn)
 
-	filteredConns, err := cs.connDumper.DumpFlows(openflow.CtZone)
+	filteredConns, totalConns, err := cs.connDumper.DumpFlows(openflow.CtZone)
```
pretty minor comment, but I am not a big fan of the naming here, since we have `filteredConns`, `totalConns` and `maxConns`, but the first one is a list while the others are integers. I would suggest renaming the first one to `filteredConnsList` or renaming the others to `totalNumConns` and `maxNumConns`.

dreamtalen replied:
Thanks, I renamed the `filteredConns` to `filteredConnsList`.

antoninbas reviewed (commented):
some small comments

antoninbas commented on pkg/agent/flowexporter/connections/interface.go:
```diff
@@ -22,5 +22,8 @@ import (
-	DumpFlows(zoneFilter uint16) ([]*flowexporter.Connection, error)
+	// DumpFlows returns a list of filtered connections and the number of total connections.
+	DumpFlows(zoneFilter uint16) ([]*flowexporter.Connection, int, error)
+	// GetMaxConnections returns the size of connection traking table.
```
```suggestion
	// GetMaxConnections returns the size of the connection tracking table.
```

dreamtalen replied:
Thanks, fixed.

antoninbas reviewed (commented):
One typo, otherwise LGTM

antoninbas commented:
/test-all

tnqn commented on pkg/agent/flowexporter/connections/connections.go:
```diff
@@ -190,16 +192,22 @@ func (cs *ConnectionStore) Poll() (int, error) {
-	connsLen := len(filteredConns)
-	filteredConns = nil
+	connsLen := len(filteredConnsList)
+	filteredConnsList = nil
```
not introduced by this PR, but why this line is needed?

srikartati replied:
Yes, not required. It was a stray line from a past debugging price. Thanks for pointing it out. Should be removed in the future.

tnqn reviewed (approved):
LGTM, a question inline, but can be addressed later.
